lib/posix/options/mmap.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024, Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stddef.h>
+#include <sys/types.h>
+
+#include <kernel_arch_interface.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/fdtable.h>
+#include <zephyr/posix/sys/mman.h>
+#include <zephyr/posix/unistd.h>
+
+#define _page_size COND_CODE_1(CONFIG_MMU, (CONFIG_MMU_PAGE_SIZE), (PAGE_SIZE))
+
+int zvfs_ioctl(int fd, int cmd, va_list args);
+
+static int p2z(int prot, int pflags)
+{
+	bool rw = (prot & PROT_WRITE) != 0;
+	bool ex = (prot & PROT_EXEC) != 0;
+	bool fixed = (pflags & MAP_FIXED) != 0;
+	bool shared = (pflags & MAP_SHARED) != 0;
+	bool private = (pflags & MAP_PRIVATE) != 0;
+
+	if (!(shared ^ private)) {
+		return -1;
+	}
+
+	return (rw * K_MEM_PERM_RW) | (ex * K_MEM_PERM_EXEC) | (fixed * K_MEM_DIRECT_MAP);
+}
+
+static inline int zvfs_ioctl_wrap(int fd, int cmd, ...)
+{
+	int ret;
+	va_list args;
+
+	va_start(args, cmd);
+	ret = zvfs_ioctl(fd, cmd, args);
+	va_end(args);
+
+	return ret;
+}
+
+void *mmap(void *addr, size_t len, int prot, int flags, int fd, off_t off)
+{
+	int zflags = p2z(prot, flags);
+
+	if ((len == 0) || (zflags == -1)) {
+		errno = EINVAL;
+		return MAP_FAILED;
+	}
+
+	if (fd > 0) {
+		/* non-anonymous mapping */
```

Comment:
Driving by, haven't had time for a full review but this all looks pretty straightforward:

Is MAP_ANON/MAP_ANONYMOUS not part of POSIX?  You probably want that here for portability, since that's the primary mechanism in linux.  Linux will ignore the fd argument if it's set (it doesn't have to be -1).  This code will use the fd only, meaning that Linux code that gets ported to it (i.e. drops the MAP_ANONYMOUS flag) and forgets to pass -1 in the fd argument (because it was never needed) is going to misbehave.

Reply:
> Driving by, haven't had time for a full review but this all looks pretty straightforward:
> 
> Is MAP_ANON/MAP_ANONYMOUS not part of POSIX? 

It is (maybe) surprisingly not POSIX.

> You probably want that here for portability, since that's the primary mechanism in linux. 

Probably, yeah.

> Linux will ignore the fd argument if it's set (it doesn't have to be -1).  This code will use the fd only, meaning that Linux code that gets ported to it (i.e. drops the MAP_ANONYMOUS flag) and forgets to pass -1 in the fd argument (because it was never needed) is going to misbehave.

I was thinking something along the line of forcing the fd argument to -1 if the `MAP_ANONYMOUS` flag is present. That should work, right?

Since this PR is already holding up quite a few more, would it be acceptable to submit that change as a follow up, @andyross?

Reply:
Oh sure; it was just a note. I try to be pretty clear with my -1's. :)

Reply:
Fixed!